Julia0.6

mfalt opened the pull request:
Work in progress.
Fixes warnings on 0.6.
Fixes abstract types #23
More work to be done before pull.

**Notable performance examples:** taken from tests, observe that first calls include allocation

ProximalOperators.IndAffine
description : indicator of an affine subspace
domain      : AbstractArray{Real,1}, AbstractArray{Complex,1}
expression  : x ↦ 0 if Ax = b, +∞ otherwise
parameters  : A = Array{Float64,2} of size (1, 10), b = Array{Float64,1} of size (1,)

OLD:

* call                        :   **0.006850 seconds (2.19 k allocations: 106.342 KiB)**
* prox        :   **0.649409 seconds (515.29 k allocations: 24.416 MiB**, 0.88% gc time)
* prox!       :   **0.000016** seconds (15 allocations: 1.125 KiB)
* prox_naive  :   0.000011 seconds (17 allocations: 1.172 KiB)
* prox        :   **0.000005** seconds (17 allocations: 1.313 KiB)
* prox!       :   **0.000004** seconds (15 allocations: 1.125 KiB)
* prox_naive  :   0.000005 seconds (17 allocations: 1.172 KiB)

NEW:

* call                        :   **0.000004 seconds (2 allocations: 192 bytes)**
* prox        :   **0.000007 seconds (17 allocations: 1.313 KiB)**
* prox!       :   **0.000001** seconds (15 allocations: 1.125 KiB)
* prox_naive  :   0.000001 seconds (16 allocations: 1.156 KiB)
* prox        :   **0.000001** seconds (17 allocations: 1.313 KiB)
* prox!       :   **0.000001** seconds (15 allocations: 1.125 KiB)
* prox_naive  :   0.000001 seconds (16 allocations: 1.156 KiB)

ProximalOperators.LeastSquares
description : least-squares penalty
domain      : AbstractArray{Real,1}
expression  : x ↦ (λ/2)||A*x - b||^2
parameters  : λ = 1.0, A = Array{Float64,2} of size (10, 25), b = Array{Float64,1} of size (10,)
OLD:

* call                        :   **0.002280 seconds (17 allocations: 1.266 KiB)**
* prox        :   **0.039139 seconds (14.41 k allocations: 705.142 KiB)**
* prox!       :   **0.000055** seconds (17 allocations: 2.391 KiB)
* prox_naive  :   0.004583 seconds (47 allocations: 19.031 KiB)
* prox        :   **0.000046 seconds (32 allocations: 4.734 KiB)**
* prox!       :   0.000007 seconds (17 allocations: 2.391 KiB)
* prox_naive  :   0.000049 seconds (22 allocations: 17.063 KiB)

NEW:
* call                        :   **0.000005 seconds (2 allocations: 320 bytes)**
* prox        :   **0.000054 seconds (20 allocations: 4.547 KiB)**
* prox!       :   **0.000004** seconds (10 allocations: 2.281 KiB)
* prox_naive  :   0.006062 seconds (14 allocations: 16.938 KiB)
* prox        :   **0.000009 seconds (20 allocations: 4.547 KiB)**
* prox!       :   0.000004 seconds (10 allocations: 2.281 KiB)
* prox_naive  :   0.000029 seconds (14 allocations: 16.938 KiB)

ProximalOperators.LeastSquares
description : least-squares penalty
domain      : AbstractArray{Real,1}
expression  : x ↦ (λ/2)||A*x - b||^2
parameters  : λ = 0.2654585810275454, A = Array{Float64,2} of size (40, 13), b = Array{Float64,1} of size (40,)
OLD:
* call                        :   0.000006 seconds (6 allocations: 960 bytes)
* prox        :   **0.000031** seconds (28 allocations: 4.922 KiB)
* prox!       :   0.000005 seconds (13 allocations: 1.563 KiB)
* prox_naive  :   0.000019 seconds (22 allocations: 6.234 KiB)
* prox        :   **0.000016** seconds (28 allocations: 4.922 KiB)
* prox!       :   0.000004 seconds (13 allocations: 1.563 KiB)
* prox_naive  :   0.000015 seconds (22 allocations: 6.234 KiB)

NEW:
* call                        :   0.000002 seconds (2 allocations: 896 bytes)
* prox        :   **0.000010** seconds (16 allocations: 4.734 KiB)
* prox!       :   0.000003 seconds (6 allocations: 1.453 KiB)
* prox_naive  :   0.000014 seconds (14 allocations: 6.109 KiB)
* prox        :   **0.000007** seconds (16 allocations: 4.734 KiB)
* prox!       :   0.000003 seconds (6 allocations: 1.453 KiB)
* prox_naive  :   0.000011 seconds (14 allocations: 6.109 KiB)

ProximalOperators.LeastSquares
description : least-squares penalty
domain      : AbstractArray{Complex,1}
expression  : x ↦ (λ/2)||A*x - b||^2
parameters  : λ = 1.0, A = Array{Complex{Float64},2} of size (25, 10), b = Array{Complex{Float64},1} of size (25,)

OLD:
* call                        :   **0.004213 seconds (2.00 k allocations: 94.550 KiB)**
* prox        :   **0.054817 seconds (15.56 k allocations: 787.318 KiB)**
* prox!       :   **0.000040** seconds (13 allocations: 1.891 KiB)
* prox_naive  :   **0.666595 seconds (450.84 k allocations: 20.933 MiB)**
* prox        :   **0.000052** seconds (28 allocations: 5.922 KiB)
* prox!       :   **0.000009** seconds (13 allocations: 1.891 KiB)
* prox_naive  :   **0.000037** seconds (22 allocations: 7.469 KiB)

NEW:
* call                        :   **0.000025 seconds (2 allocations: 1.063 KiB)**
* prox        :   **0.000071 seconds (16 allocations: 5.734 KiB)**
* prox!       :   **0.000005** seconds (6 allocations: 1.781 KiB)
* prox_naive  :   **0.000034 seconds (14 allocations: 7.344 KiB)**
* prox        :   **0.000022** seconds (16 allocations: 5.734 KiB)
* prox!       :   **0.000003** seconds (6 allocations: 1.781 KiB)
* prox_naive  :   **0.000020** seconds (14 allocations: 7.344 KiB)

----------------------------------------------------------
ProximalOperators.LeastSquares
description : least-squares penalty
domain      : AbstractArray{Real,1}
expression  : x ↦ (λ/2)||A*x - b||^2
parameters  : λ = 0.3611825580251262, A = SparseMatrixCSC{Float64,Int64} of size (100, 1000), b = Array{Float64,1} of size (100,)

OLD:
* call                        :   **0.037073 seconds (11.04 k allocations: 566.955 KiB)**
* prox        :   **0.330995** seconds (100.69 k allocations: 5.369 MiB)
* prox!       :   0.000107 seconds (27 allocations: 46.039 KiB)
* prox_naive  :   **0.732980** seconds (390.98 k allocations: 51.149 MiB, 2.35% gc time)
* prox        :   0.000703 seconds (95 allocations: 830.305 KiB)
* prox!       :   0.000107 seconds (27 allocations: 46.039 KiB)
* prox_naive  :   0.109704 seconds (5.14 k allocations: 33.245 MiB, 17.00% gc time)

NEW:
* call                        :   **0.000046 seconds (2 allocations: 1.750 KiB)**
* prox        :   **0.134826** seconds (24.31 k allocations: 1.987 MiB)
* prox!       :   0.000098 seconds (19 allocations: 45.914 KiB)
* prox_naive  :   **0.243142** seconds (38.70 k allocations: 34.512 MiB, 2.21% gc time)
* prox        :   0.000713 seconds (83 allocations: 826.703 KiB)
* prox!       :   0.000104 seconds (19 allocations: 45.914 KiB)
* prox_naive  :   0.109826 seconds (5.15 k allocations: 33.026 MiB, 20.42% gc time)

mfalt committed:
Last warnings on 0.6

mfalt committed:
Type stability and 0.6

mfalt commented on src/functions/leastSquares.jl:
```diff
@@ -27,27 +27,44 @@ type LeastSquares{R <: RealOrComplex} <: ProximableConvex
+  function LeastSquares{RC,R,M,V}(A::M, b::V, lambda::R) where {RC <: AbstractArray, R<:Real, M<:AbstractArray{RC,2}, V<:AbstractArray{RC,1}}
+    warn("Could not infer type of Factorization for $M in LeastSquares, this type will be type-unstable")
+    LeastSquares{RC,R,M,V,Factorization}(A,b,lambda)
+  end
```
This is definitely quite ugly, but does improve performance by having type stable constructors and making sure the types are not abstract.

mfalt replied:
These are all inner constructors for convenience, is there a better way to do this?

mfalt commented on src/functions/indAffine.jl:
```diff
@@ -26,11 +26,11 @@ immutable IndAffine{T <: RealOrComplex} <: IndicatorConvex
-IndAffine{T <: RealOrComplex}(A::AbstractArray{T,2}, b::AbstractArray{T,1}) =
-  IndAffine{T}(A, b)
+IndAffine{T <: RealOrComplex, M<:AbstractArray{T,2}, V<:AbstractArray{T,1}}(A::M, b::V) =
+  IndAffine{T,M,V}(A, b)
```
Should this function really needed?

lostella replied:
I used some of these outer constructors to avoid the following situation:
```julia
julia> immutable MyType{R <: Real}
           a::R
           function MyType(a::R)
               if a <= 0 error("a must be positive") end
           end
       end

julia> b = MyType(Float64(5.0))
ERROR: MethodError: Cannot `convert` an object of type Float64 to an object of type MyType{R<:Real}
This may have arisen from a call to the constructor MyType{R<:Real}(...),
since type constructors fall back to convert methods.
 in MyType{R<:Real}(::Float64) at ./sysimg.jl:53

julia> b = MyType{Float64}(Float64(5.0)) # this one is OK but ugly to use

julia> MyType{R <: Real}(a::R) = MyType{R}(a) # define outer constructor
MyType{R<:Real}

julia> b = MyType(Float64(5.0)) # now this is also OK
``` 
Is there a better way to do this? Has something changed wrt this in 0.6?

mfalt replied:
No I think you are right, but then we should be able to get rid of several of the inner constructors right? But maybe we want to keep them for some reason?

lostella replied:
It might be that we can in some cases. I simply followed this logic: outer constructors are to make things convenient, i.e., to allow not specifying types, to default some construction arguments, to allow for different ways of specifying the arguments (indAffine.jl is a good example of this, see the second outer construtor there); inner constructors is where all necessary checks, factorizations, normalizations occur.

mfalt replied:
I see, the problem with inner constructors is that we need to know all the parameters in the type when the constructor is called. I'm not sure how this is possible, for example `F` in the case with `LeastSquares` if the factorization is done in the inner constructor (or actually not done in the constructors at all).

mfalt committed:
Push travis version

lostella commented:
Check out also #27, you may want to merge it into your branch since it touches almost all functions.

mfalt committed:
Merge remote lostella/less-types

lostella commented:
The speedups look impressive! We should probably set up a bunch of fixed set of cases to use as benchmarks, at least for the most important functions, and measure the time required to `prox!' using `@benchmark' and other utilities from BenchmarkTools.jl.

mfalt commented:
Yes, I agree, they have something similar for the julia repo, we might want
to take a look at that.

On Sat, Apr 8, 2017, 22:17 Lorenzo Stella <notifications@github.com> wrote:

> The speedups look impressive! We should probably set up a bunch of fixed
> set of cases to use as benchmarks, at least for the most important
> functions, and measure the time required to prox!' using@benchmark
> <https://github.com/benchmark>' and other utilities from
> BenchmarkTools.jl.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kul-forbes/ProximalOperators.jl/pull/26#issuecomment-292742935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADFlSVIqI16heEbew-KXkRKDd0dPWx1fks5rt-s_gaJpZM4M3gfr>
> .
>

mfalt committed:
Julia0.6mfalt (#28)

* Sparse indaffine

* Minor bugs

mfalt committed:
Added sparse IndAffine tests

mfalt committed:
Appveyor julia bump

lostella committed:
Merge branch 'master' into julia0.6

lostella commented:
I have opened #35 to propose some changes to `SeparableSum` and `SlicedSeparableSum`. If we all agree on those changes we can probably merge #35, then resolve all conflicts here and probably merge also #26?

lostella commented:
The other option is that I push those modifications directly on this branch as new commits.

nantonel commented:
I'm in favour of the last option:

> EDIT: or maybe, we first merge #26 and then verify how #35 fits into that. Maybe this is better.

mfalt committed:
Avoid sparse inverse broadcast

Avoids doing sparse broadcast with division which results in full sparse matrices

lostella committed:
removed warning

lostella committed:
Merge branch 'master' into julia0.6
